File: python_test/algorithm/vertical_order_optimizer.py
"""
Phase 4: Vertical Order Optimizer

Optimizes vertical order within each layer to minimize edge crossings.
Groups entities by their targets in the next layer.
"""

from typing import List, Tuple
from collections import defaultdict


class VerticalOrderOptimizer:
    """
    Optimizes the vertical order (Y-axis) of entities within each layer
    to minimize edge crossings by grouping entities with same targets.
    """

    def __init__(self, relations: List[Tuple[str, str]]):
        self.relations = relations

    def optimize(
        self,
        horizontal_layers: List[List[str]],
        entity_order: List[str]
    ) -> List[List[str]]:
        """
        Optimize vertical order within each layer

        Args:
            horizontal_layers: Initial layer assignments from horizontal classifier
            entity_order: Global entity ordering (for tie-breaking)

        Returns: Optimized layers with reordered entities
        """
        optimized_layers = [layer[:] for layer in horizontal_layers]

        # Process layers from right to left (start with last layer)
        # Last layer: order by entity_order
        last_idx = len(optimized_layers) - 1
        if last_idx >= 0:
            last_layer = optimized_layers[last_idx]
            ordered_last = [e for e in entity_order if e in last_layer]
            # Add any remaining (shouldn't happen, but safety)
            for e in last_layer:
                if e not in ordered_last:
                    ordered_last.append(e)
            optimized_layers[last_idx] = ordered_last

        # Other layers: sort by targets in next layer
        for layer_idx in range(len(optimized_layers) - 2, -1, -1):
            current_layer = optimized_layers[layer_idx]
            next_layer = optimized_layers[layer_idx + 1]

            reordered = self._sort_layer_by_targets(
                current_layer,
                next_layer,
                entity_order
            )
            optimized_layers[layer_idx] = reordered

        return optimized_layers

    def _sort_layer_by_targets(
        self,
        current_layer: List[str],
        next_layer: List[str],
        entity_order: List[str]
    ) -> List[str]:
        """
        Sort layer by grouping entities with same targets in next layer

        Algorithm:
        1. Find what each entity points to in next layer
        2. Determine primary target (first in next_layer order)
        3. Group entities by primary target
        4. Order groups by target position in next_layer
        5. Within each group, sort by entity_order
        """
        # Find targets for each entity in current layer
        entity_to_targets = {}
        for entity in current_layer:
            targets = [
                right for left, right in self.relations
                if left == entity and right in next_layer
            ]
            entity_to_targets[entity] = targets

        # Determine primary target for each entity
        # Primary target = first target in next_layer order
        entity_to_primary_target = {}
        for entity, targets in entity_to_targets.items():
            if targets:
                # Primary target is the first one appearing in next_layer
                primary = min(targets, key=lambda t: next_layer.index(t))
                entity_to_primary_target[entity] = primary
            else:
                # No target: assign None (will be placed last)
                entity_to_primary_target[entity] = None

        # Group entities by primary target
        target_groups = defaultdict(list)
        for entity in current_layer:
            primary = entity_to_primary_target[entity]
            target_groups[primary].append(entity)

        # Sort each group by entity_order
        for target in target_groups:
            target_groups[target].sort(key=lambda e: entity_order.index(e) if e in entity_order else float('inf'))

        # Build final order: groups ordered by target position in next_layer
        reordered = []

        # First: entities with targets (ordered by target position)
        for target in next_layer:
            if target in target_groups:
                reordered.extend(target_groups[target])

        # Last: entities with no targets
        if None in target_groups:
            reordered.extend(target_groups[None])

        return reordered

File: src/infrastructure/layout/phases/VerticalOrderOptimizer.ts
/**
 * Phase 4: Vertical Order Optimizer
 *
 * Optimizes vertical order within each layer to minimize edge crossings.
 * Groups entities by their targets in the next layer.
 */

import { DirectedRelation } from './types';

export class VerticalOrderOptimizer {
  /**
   * Optimize vertical order (Y-axis) within each layer
   *
   * Algorithm:
   * 1. Process layers right to left (start with last layer)
   * 2. Last layer: order by entityOrder
   * 3. Other layers: group entities by targets in next layer
   * 4. Sort groups to minimize crossings
   *
   * @param horizontalLayers - Initial layer assignments
   * @param entityOrder - Global entity ordering (for tie-breaking)
   * @param relations - Directed relations
   * @returns Optimized layers with reordered entities
   */
  static optimize(
    horizontalLayers: string[][],
    entityOrder: string[],
    relations: DirectedRelation[]
  ): string[][] {
    console.log('\n=== PHASE 4: VERTICAL ALIGNMENT (Y-AXIS) ===');

    if (horizontalLayers.length === 0) return horizontalLayers;

    // Deep copy layers
    const layers = horizontalLayers.map(layer => [...layer]);

    // Last layer: order by entity_order
    const lastLayerIdx = layers.length - 1;
    const lastLayer = layers[lastLayerIdx];

    const orderedLast: string[] = [];
    for (const entity of entityOrder) {
      if (lastLayer.includes(entity)) {
        orderedLast.push(entity);
      }
    }

    // Add any remaining (shouldn't happen, but safety)
    for (const entity of lastLayer) {
      if (!orderedLast.includes(entity)) {
        orderedLast.push(entity);
      }
    }

    layers[lastLayerIdx] = orderedLast;

    // Other layers: sort by targets in next layer (right to left)
    for (let layerIdx = layers.length - 2; layerIdx >= 0; layerIdx--) {
      const currentLayer = layers[layerIdx];
      const nextLayer = layers[layerIdx + 1];

      const reordered = this.sortLayerByTargets(
        currentLayer,
        nextLayer,
        entityOrder,
        relations
      );

      layers[layerIdx] = reordered;
    }

    console.log('Vertical optimization complete');

    return layers;
  }

  /**
   * Sort layer by grouping entities with same targets in next layer
   *
   * Algorithm:
   * 1. Find what each entity points to in next layer
   * 2. Determine primary target (first in next_layer order)
   * 3. Group entities by primary target
   * 4. Order groups by target position in next_layer
   * 5. Within each group, sort by entity_order
   */
  private static sortLayerByTargets(
    currentLayer: string[],
    nextLayer: string[],
    entityOrder: string[],
    relations: DirectedRelation[]
  ): string[] {
    // Find targets for each entity in current layer
    const entityToAllTargets = new Map<string, string[]>();
    for (const entity of currentLayer) {
      const targets: string[] = [];
      for (const { left, right } of relations) {
        if (left === entity && nextLayer.includes(right)) {
          targets.push(right);
        }
      }
      entityToAllTargets.set(entity, targets);
    }

    // Determine primary target for each entity
    // Primary target = first target in next_layer order
    const entityPrimaryTarget = new Map<string, string | null>();
    for (const entity of currentLayer) {
      const targets = entityToAllTargets.get(entity) || [];
      if (targets.length > 0) {
        // Take the first target (or the one with min position)
        const primary = targets.reduce((best, current) => {
          return nextLayer.indexOf(current) < nextLayer.indexOf(best) ? current : best;
        });
        entityPrimaryTarget.set(entity, primary);
      } else {
        entityPrimaryTarget.set(entity, null);
      }
    }

    // Group by primary target
    const targetGroups = new Map<string | null, string[]>();
    for (const entity of currentLayer) {
      const primary = entityPrimaryTarget.get(entity)!;
      if (!targetGroups.has(primary)) {
        targetGroups.set(primary, []);
      }
      targetGroups.get(primary)!.push(entity);
    }

    // Sort each group by entity_order
    for (const [target, group] of targetGroups.entries()) {
      targetGroups.set(
        target,
        group.sort((a, b) => {
          const indexA = entityOrder.indexOf(a);
          const indexB = entityOrder.indexOf(b);
          if (indexA === -1 && indexB === -1) return 0;
          if (indexA === -1) return 1;
          if (indexB === -1) return -1;
          return indexA - indexB;
        })
      );
    }

    // Order groups by target position in nextLayer
    const orderedTargets = Array.from(targetGroups.keys()).sort((a, b) => {
      if (a === null) return 1; // Entities with no target go last
      if (b === null) return -1;
      return nextLayer.indexOf(a) - nextLayer.indexOf(b);
    });

    // Build final ordered layer
    const orderedLayer: string[] = [];
    for (const target of orderedTargets) {
      orderedLayer.push(...targetGroups.get(target)!);
    }

    return orderedLayer;
  }
}
